bootstrap/src/main/java/jenkins/bootstrap/Bootstrap.java

```diff
@@ -0,0 +1,313 @@
+ * all copies or substantial portions of the Software.
+ * 
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package jenkins.bootstrap;
+
+import javax.naming.Context;
+import javax.naming.InitialContext;
+import javax.naming.NamingException;
+import javax.servlet.ServletContext;
+import javax.servlet.ServletContextEvent;
+import javax.servlet.ServletContextListener;
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.net.URL;
+import java.net.URLClassLoader;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.Date;
+import java.util.List;
+import java.util.ServiceLoader;
+import java.util.Set;
+import java.util.logging.Logger;
+
+import static java.util.logging.Level.*;
+
+/**
+ * Entry point when Jenkins is used as a webapp.
+ *
+ * @author Kohsuke Kawaguchi
+ */
+public class Bootstrap implements ServletContextListener {
+
+    private List<BootLogic> bootLogics = Collections.emptyList();
+    private ServletContext context;
+    private ClassLoader coreClassLoader;
+    /**
+     * JENKINS_HOME
+     */
+    private File home;
+
+    /**
+     * Creates the sole instance of {@link jenkins.model.Jenkins} and register it to the {@link ServletContext}.
+     */
+    public void contextInitialized(ServletContextEvent event) {
+        context = event.getServletContext();
+        context.setAttribute(Bootstrap.class.getName(),this);   // allow discovery of this
+        try {
+
+            final FileAndDescription describedHomeDir = getHomeDir(event);
+            home = describedHomeDir.file.getAbsoluteFile();
+            home.mkdirs();
+            System.out.println("Jenkins home directory: "+ home +" found at: "+describedHomeDir.description);
+
+            // check that home exists (as mkdirs could have failed silently), otherwise throw a meaningful error
+            if (!home.exists())
+                throw new Error("Invalid JENKINS_HOME: "+ home);
+
+            recordBootAttempt(home);
+
+            coreClassLoader = buildCoreClassLoader();
+            Thread.currentThread().setContextClassLoader(coreClassLoader);
+
+            bootLogics = loadBootLogics(coreClassLoader);
+
+            for (BootLogic b : bootLogics) {
+                b.contextInitialized(event);
+            }
+        } catch (Error | RuntimeException e) {
+            LOGGER.log(SEVERE, "Failed to initialize Jenkins",e);
+            throw e;
+        } catch (Exception e) {
```

Comment:
This kind of handlers will miss Types, which inherit from `Throwable` directly. And we have them in Jenkins: https://github.com/search?p=1&q=org%3Ajenkinsci+%22extends+Throwable%22&type=Code